server/src/main/java/org/apache/iotdb/db/query/dataset/NonAlignEngineDataSet.java

```diff
@@ -0,0 +1,354 @@
+                      }
+                      ReadWriteIOUtils.write(longValue, valueBAOS);
+                      break;
+                    case FLOAT:
+                      float floatValue = batchData.getFloat();
+                      if (encoder != null && encoder.needEncode(time)) {
+                        floatValue = encoder.encodeFloat(floatValue, time);
+                      }
+                      ReadWriteIOUtils.write(floatValue, valueBAOS);
+                      break;
+                    case DOUBLE:
+                      double doubleValue = batchData.getDouble();
+                      if (encoder != null && encoder.needEncode(time)) {
+                        doubleValue = encoder.encodeDouble(doubleValue, time);
+                      }
+                      ReadWriteIOUtils.write(doubleValue, valueBAOS);
+                      break;
+                    case BOOLEAN:
+                      ReadWriteIOUtils.write(batchData.getBoolean(),
+                              valueBAOS);
+                      break;
+                    case TEXT:
+                      ReadWriteIOUtils
+                              .write(batchData.getBinary(),
+                                      valueBAOS);
+                      break;
+                    default:
+                      throw new UnSupportedDataTypeException(
+                              String.format("Data type %s is not supported.", type));
+                  }
+                }
+                batchData.next();
+              }
+              else {
+                if (reader.hasNextBatch()) {
+                  batchData = reader.nextBatch();
+                  dataSet.cachedBatchData[index] = batchData;
+                  continue;
+                }
+                else
+                  break;
+              }
+              if (dataSet.offsetArray.get(index) == 0) {
+                rowCount++;
+                if (dataSet.limit > 0) {
+                  dataSet.alreadyReturnedRowNumArray.incrementAndGet(index);
+                }
+              } else {
+                dataSet.offsetArray.decrementAndGet(index);
+              }
+            }
+            if (rowCount == 0) {
+              blockingQueue.put(new Pair(null, null));
+              // set the hasRemaining field in reader to false
+              // tell the Consumer not to submit another task for this reader any more
+              reader.setHasRemaining(false);
+              // remove itself from the QueryTaskPoolManager
+              reader.setManagedByQueryManager(false);
+              return;
+            }
+
+            ByteBuffer timeBuffer = ByteBuffer.allocate(timeBAOS.size());
+            timeBuffer.put(timeBAOS.getBuf(), 0, timeBAOS.size());
+            timeBuffer.flip();
+            ByteBuffer valueBuffer = ByteBuffer.allocate(valueBAOS.size());
+            valueBuffer.put(valueBAOS.getBuf(), 0, valueBAOS.size());
+            valueBuffer.flip();
```

Comment:
You may just wrap the byte[] in the BAOS with `ByteBuffer.wrap()` as it does not seem you will reuse it.

Reply:
fixed